common/nodejs-utils/src/bin/update_node_inventory.rs

```diff
@@ -0,0 +1,168 @@
+// Required due to: https://github.com/rust-lang/rust/issues/95513
+#![allow(unused_crate_dependencies)]
+
+use anyhow::{Context, Result};
+use heroku_inventory_utils::{
+    checksum::Checksum,
+    inv::{read_inventory_file, Arch, Artifact, Inventory, Os},
+};
+use node_semver::Version;
+use serde::Deserialize;
+use sha2::Sha256;
+use std::{
+    collections::{HashMap, HashSet},
+    env, fs, process,
+};
+
+/// Updates the local node.js inventory.toml with versions published on nodejs.org.
+fn main() {
+    let inventory_path = env::args().nth(1).unwrap_or_else(|| {
+        eprintln!("Usage: update_inventory <path/to/inventory.toml>");
+        process::exit(2);
+    });
+
+    let inventory_artifacts: HashSet<Artifact<Version, Sha256>> =
+        read_inventory_file(&inventory_path)
+            .unwrap_or_else(|e| {
+                eprintln!("Error reading inventory at '{inventory_path}': {e}");
+                std::process::exit(1);
+            })
+            .artifacts
+            .into_iter()
+            .collect();
+
+    let upstream_artifacts = list_upstream_artifacts().unwrap_or_else(|e| {
+        eprintln!("Failed to fetch upstream node.js versions: {e}");
+        process::exit(4);
+    });
+
+    let inventory = Inventory {
+        artifacts: upstream_artifacts,
+    };
+
+    let toml = toml::to_string(&inventory).unwrap_or_else(|e| {
+        eprintln!("Error serializing inventory as toml: {e}");
+        process::exit(6);
+    });
+
+    fs::write(inventory_path, toml).unwrap_or_else(|e| {
+        eprintln!("Error writing inventory file: {e}");
+        process::exit(7);
+    });
+
+    let remote_artifacts: HashSet<Artifact<Version, Sha256>> =
+        inventory.artifacts.into_iter().collect();
+
+    [
+        ("Added", &remote_artifacts - &inventory_artifacts),
+        ("Removed", &inventory_artifacts - &remote_artifacts),
+    ]
+    .iter()
+    .filter(|(_, artifact_diff)| !artifact_diff.is_empty())
+    .for_each(|(action, artifacts)| {
+        let mut list: Vec<&Artifact<Version, Sha256>> = artifacts.iter().collect();
+        list.sort_by_key(|a| &a.version);
+        println!(
+            "{} {}.",
+            action,
+            list.iter()
+                .map(ToString::to_string)
+                .collect::<Vec<_>>()
+                .join(", ")
+        );
+    });
+}
+
+fn list_upstream_artifacts() -> Result<Vec<Artifact<Version, Sha256>>, anyhow::Error> {
+    let target_version = Version::parse("0.8.6").context("Failed to parse version")?;
+
+    list_releases()?
+        .into_iter()
+        .filter(|release| release.version >= target_version)
```

Comment:
It would be nice if this accounted for versions already in the inventory so that we don't have to make a ton of requests for `SHASUMS256.txt` files.

Reply:
I took a stab at this here: https://github.com/heroku/buildpacks-nodejs/compare/4d5f13b03a25900a1844103281f6f5570b538d3c..7e46492f0403d9bf9d0a5bcb033ed69e80675867 -- basically inlined the functions (to avoid having to pass around the list of local/inventory artifacts), and cloning artifacts when there's one in the inventory that matches the os, arch and version for an upstream release.

Not sure it's super elegant, but readability isn't much worse than before :-) And it completes in 1-2 seconds now (vs about a minute with the previous approach)

Reply:
Just gave it a quick run-through. Seems to work well and I don't find it unreadable 😅